### PR TITLE
GODMODE V2

### PR DESCRIPTION
**Pull request by J0hann5**
These changes allow for a working predator-prey simulation. 

## Changes:
- Entity standard values
- Entity spawning is now random
- Entity directory is now /data/entities
- Increased Chart Performance (+ custom Setting)
- Increased Map Performance (+ custom Setting)
- Reworked eat function 
- Impact of child value

### CODE REVIEW

File: credits/index.html [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            <p>Originally made by <a href="https://github.com/konrad2002">Konrad2002</a>.</p>
+            <p>Modified by Jonas and Johann to get the predator-prey interactions working as part of a Biology task.</p><br>
+            <p>Graph made using <a href="https://www.highcharts.com/">Highcharts</a>.</p>
+            <p>This project uses <a href="https://jquery.com/">jQuery</a>.</p>
+            </div>
+    
+    </div>
+</body>
+
+</html>
```
Comment (J0hann5) [inline credits/index.html:43]: Added Credits

File: data/entities/1000/ms.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+        "neededMass": 70,
+        "speed": 50,
+        "strength": 15,
+        "consumes": [6, 14],
+        "reproduction": {
+            "type": "animal",
+            "childs": 2
+        }
+    }
+}
```
Comment (J0hann5) [inline data/entities/1000/ms.json:19]: New entity file location

File: credits/index.html (hunk 1 above)
Comment (J0hann5) [inline credits/index.html]: Added Credits

File: data/entities/ms.json [hunk 3]
```diff
@@ -3,16 +3,17 @@
...
+        "speed": 50,
+        "strength": 15,
+        "consumes": [6, 14],
         "reproduction": {
             "type": "animal",
-            "childs": 0.2
+            "childs": 2
         }
     }
 }
```
Comment (J0hann5) [inline data/entities/ms.json:19]: - Removed unnecessary Values
- Tweaked values to make preditor-prey work

File: editor/index.php [hunk 4]
```diff
@@ -87,7 +87,7 @@
 
                     foreach ($types as $type) {
                         echo("<textarea name='new".$type."' cols='50' rows='21'>");
-                        echo(file_get_contents($_SERVER['DOCUMENT_ROOT']."/api/godmode/data/entities/".$pid."/".$type.".json"));
+                        echo(file_get_contents($_SERVER['DOCUMENT_ROOT']."/godmode/data/entities/".$pid."/".$type.".json"));
                         echo("</textarea>");
                     }
```
Comment (J0hann5) [inline editor/index.php:93]: tweaked paths

File: index.html [hunk 5]
```diff
@@ -10,6 +10,7 @@
 
     <script src="https://code.highcharts.com/highcharts.js"></script>
     <script src="https://code.highcharts.com/modules/data.js"></script>
+    <script src="https://code.highcharts.com/modules/boost.js"></script>
```
Comment (J0hann5) [inline index.html:13]: Highcharts Performance boost Module

File: index.html [hunk 6]
```diff
@@ -60,42 +61,44 @@ <h2>Configs</h2><br>
...
                 </div>
 
                 <div class="config-box">
-                    <h3>Properties</h3>
-                    <input type="number" id="propertyId" value="1000">
-                    <textarea id="propertiesText" rows="15" cols="25"></textarea>
+                    <h3>Controls</h3><br>
+                    <button type="button" id="newPopulationBtn">New Population</button><br><br>
+                    <!-- <button type="button" id="doStepBtn">Step</button> -->
+                    <button type="button" id="startLoopBtn">Toggle Loop</button>
```
Comment (J0hann5) [inline index.html:101]: new Settings and starting values

File: js/chart.js [hunk 7]
```diff
@@ -151,10 +152,12 @@ function drawChart(main) {
                 valueDecimals: dataset
             },
             series: [{
+                animation: false,
                 data: dataset,
                 name: main.populations[i].type,
                 type: 'line',
-                color: Highcharts.getOptions().colors[i],
+                lineWidth: 4,
+                color: Highcharts.getOptions().colors[i+2],
```
Comment (J0hann5) [inline js/chart.js:160]: Increases Highcharts readability and disables drawing animation

File: js/config.js [hunk 8]
```diff
@@ -1,5 +1,5 @@
 function getJson(entity, propertyId, toTextArea = false, storeResult = false) {
-    var url = 'https://logilutions.de/api/godmode/data/entities/' + propertyId + '/' + entity + '.json';
+    var url = '/godmode/data/entities/' + propertyId + '/' + entity + '.json';
```
Comment (J0hann5) [inline js/config.js:2]: tweaked directory

File: js/entity.js [hunk 9]
```diff
@@ -1,17 +1,10 @@
...
-                y: getRandomInt(0, setup.height - 1)
-            }
-        } else {
-            this.pos = {
-                x: x,
-                y: y
-            }
+        this.pos = {
+            x: x,
+            y: y
```
Comment (J0hann5) [inline js/entity.js:7]: enable random spawning

File: js/entity.js [hunk 10]
```diff
@@ -27,13 +20,18 @@ class Entity {
...
         var r = getRandomInt(0, 100);
 
         if (r > 0 && r <= this.data.properties.speed) {
             this.newDirection = getRandomInt(0, 3);
 
+            var index = main.coords[this.pos.x][this.pos.y].indexOf(this);
+            if (index !== -1) {
+                main.coords[this.pos.x][this.pos.y].splice(index, 1);
+            }
+
```
Comment (J0hann5) [inline js/entity.js:34]: new coords system

File: js/entity.js [hunk 11]
```diff
@@ -47,82 +45,80 @@ class Entity {
...
             }
 
+            main.coords[this.pos.x][this.pos.y].push(this);
+
+        }
+        if (this.data.rank == "secondary consumer") {
+            this.food -= this.consumption/2;
+        } else {
+            this.food -= this.consumption;
         }
```
Comment (J0hann5) [inline js/entity.js:55]: tweaks consumption of secondary consumer to make it survive longer without increasing needed mass

File: js/entity.js [hunk 12]
```diff
@@ -47,82 +45,80 @@ class Entity {
...
+            )
+        ) {
             if (main.populations[this.data.id - 1].entities.length < 2) {
-                console.warn("can't reproduct because of missing partner");
+                console.warn(this.data.name + " can't reproduct because a lack of partners");
             } else {
-                main.populations[this.data.id - 1].addEntity(this.pos.x + offsetX, this.pos.y + offsetY);
-                this.food -= this.data.properties.neededMass / 2;
+                main.populations[this.data.id - 1].addEntity();
+                this.food -= this.data.properties.neededMass / 2
```
Comment (J0hann5) [inline js/entity.js:73]: updated reproduction function to make child value more intuitive and match coords system

removed grammar issues

File: js/entity.js [hunk 13]
```diff
@@ -47,82 +45,80 @@ class Entity {
...
-                        if (entity.data.properties.reproduction.type == "plant" && this.data.properties.reproduction.type == "plant" && entity.dead == false && this.dead == false) {
-                            console.log("killed plant");
-                            entity.dead = true;
-                        }
                     }
-                }
-
-            });
+                })
+                // }
```
Comment (J0hann5) [inline js/entity.js:113]: new and better eat function

File: js/entity.js [hunk 14]
```diff
@@ -47,82 +45,80 @@ class Entity {
...
         });
     }
 
     checkHealth() {
 
         if (this.food <= 0 && this.data.properties.reproduction.type == "animal") {
             this.dead = true;
-            console.log(this.data.name + " died: too less food");
+            console.log(this.data.name + " died: too little food");
```
Comment (J0hann5) [inline js/entity.js:121]: wrong grammar

File: js/game.js [hunk 15]
```diff
@@ -6,9 +6,16 @@ class Game {
...
+                    var index = main.coords[entity.pos.x][entity.pos.y].indexOf(entity);
+                    if (index !== -1) {
+                        main.coords[entity.pos.x][entity.pos.y].splice(index, 1);
+                    }
+                    //console.log("Entity died: " + entity.uuid);
+                    population.entities.splice(i, 1);
+                }
 
-                entity.move(main);
+                entity.move();
```
Comment (J0hann5) [inline js/game.js:18]: coords system implemention

File: js/main.js [hunk 16]
```diff
@@ -78,7 +91,7 @@ class Main {
             i++;
         });
 
-        if (main.steps % 50 == 0) {
+        if (main.steps % this.main.setup.graphrate == 0) {
```
Comment (J0hann5) [inline js/main.js:94]: custom tweakable map and graph rate settings

File: js/main.js [hunk 17]
```diff
@@ -19,6 +20,14 @@ class Main {
...
     newPopulation() {
         this.populations = [];
+        this.coords = [];
+
+        for (let i = 0; i < this.setup.width; i++) {
+            this.coords[i] = [];
+            for (let j = 0; j < this.setup.height; j++) {
+                this.coords[i][j] = [];
+            }
+        }
```
Comment (J0hann5) [inline js/main.js:30]: coords implememtation

File: js/population.js [hunk 18]
```diff
@@ -4,14 +4,22 @@ class Population {
...
+
+        if (x === null || y === null) {
+            x = getRandomInt(0, main.setup.width - 1);
+            y = getRandomInt(0, main.setup.height - 1);
+        }
+
+        var entity = new Entity(this.type, main.setup, x, y);
+        this.entities.push(entity);
+        main.coords[x][y].push(entity);
     }
```
Comment (J0hann5) [inline js/population.js:24]: random spawning

File: js/setup.js [hunk 19]
```diff
@@ -3,6 +3,8 @@ class Setup {
         this.width = $("#width").val();
         this.height = $("#height").val();
         this.speed = $("#speed").val();
+        this.maprate = $("#maprate").val();
+        this.graphrate = $("#graphrate").val();
```
Comment (J0hann5) [inline js/setup.js:7]: new settings

File: credits/index.html [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+                    <li><a href="/godmode/editor">Editor</a></li>
+                </ul>
+            </nav>
+        </header>
+        <div class="content">
+            <h1>GodMode</h1>
+
+            <hr>
+            <h2>Credits</h2><br>
+            <p>Originally made by <a href="https://github.com/konrad2002">Konrad2002</a>.</p>
```
Comment (konrad2002) [inline credits/index.html]: and Logilutions

File: js/game.js (hunk 15 above)
Comment (konrad2002) [reply js/game.js:18]: meinz

File: js/entity.js (hunk 13 above)
Comment (konrad2002) [reply js/entity.js:113]: auch meinz

**Review by konrad2002 — APPROVED** [review]

**Pull request merged**